### PR TITLE
Checkpointing Beta

### PR DESCRIPTION
**Pull request by cekefun**
Please write a review with problems.
I will address them, but during the programming, I didn't have any better ideas.

Thank you.

### CODE REVIEW

File: src/main/cpp/checkpoint/CheckPoint.cpp [hunk 1]
```diff
@@ -205,6 +205,8 @@ void CheckPoint::WriteFileDSet(const std::string& filename, const std::string& s
 	if (!is_regular_file(fullpath)) {
 		FATAL_ERROR("Unable to find file: " + fullpath.string());
 	}
+	std::string extension = filep.extension().string();
+	std::cout << extension << std::endl;
```
Comment (Sibert-Aerts) [inline src/main/cpp/checkpoint/CheckPoint.cpp]: Maybe remove this cout

File: src/main/cpp/sim/run_stride.cpp [hunk 2]
```diff
@@ -31,6 +32,10 @@
 #include <string>
 #include <utility>
 
+std::atomic<bool> stride::util::INTERRUPT(false);
```
Comment (jonathanvdc) [inline src/main/cpp/sim/run_stride.cpp]: Can you move the definition for `stride::util::INTERRUPT` into `Signals.cpp`, remove its declaration from `ExternalVars.h` and expose it only through a function in `Signals.cpp`? That encapsulate this implementation detail better.

File: src/main/cpp/sim/run_stride.cpp [hunk 3]
```diff
@@ -31,6 +32,10 @@
...
+std::atomic<unsigned int> stride::util::INTERVAL(1);
```
Comment (jonathanvdc) [inline src/main/cpp/sim/run_stride.cpp]: I think `INTERVAL` should really be a member of the configuration, not a global variable.

File: src/main/cpp/sim/run_stride.cpp [hunk 4]
```diff
@@ -31,6 +32,10 @@
...
+std::atomic<bool> stride::util::HDF5(false);
```
Comment (jonathanvdc) [inline src/main/cpp/sim/run_stride.cpp]: Again, `HDF5` belongs in the configuration, I suppose.

File: src/main/cpp/sim/run_stride.cpp [hunk 5]
```diff
@@ -42,34 +47,54 @@ using namespace std::chrono;
...
-	// saves the start configuration
-	/*
-	if (day == 0) {
-		cp->OpenFile();
-		cp->SaveCheckPoint(sim);
-		cp->CloseFile();
+
+#if USE_HDF5
+	if (stride::util::HDF5) {
+		if (load and day == 0) {
```
Comment (jonathanvdc) [inline src/main/cpp/sim/run_stride.cpp]: These keyword aliases for `&&`, `||` and `!` are cute, but not all C++ compilers support them (Visual C++ is a notable example), and they don't appear anywhere else in our code. Can you perform the following replacements?

```cpp
and -> &&
or -> ||
not -> !
```

File: src/main/cpp/util/Signals.h [hunk 6]
```diff
@@ -1,13 +1,21 @@
 #ifndef SIGNALS_H_INCLUDED
 #define SIGNALS_H_INCLUDED
 
+#include <atomic>
```
Comment (jonathanvdc) [inline src/main/cpp/util/Signals.h]: You don't need to `#include <atomic>` here.

File: src/main/cpp/checkpoint/CheckPoint.cpp [hunk 7]
```diff
@@ -686,12 +727,221 @@ void CheckPoint::WriteExpatriates(multiregion::ExpatriateJournal& journal, boost
...
+	H5Sget_simple_extent_dims(dspace, &dims, nullptr);
+
+	std::vector<unsigned int> data(dims);
+
+	H5Dread(dset, H5T_NATIVE_UINT, H5S_ALL, H5S_ALL, H5P_DEFAULT, data.data());
+
+	H5Sclose(dspace);
+	H5Dclose(dset);
+
+	for (auto& i : data) {
```
Comment (jonathanvdc) [inline src/main/cpp/checkpoint/CheckPoint.cpp]: I don't think that what this loop does is okay. I have two main concerns:

  1. It looks for expatriates (i.e., people who are _not_ present in the population) in the population. So `p.GetId() == i` should *never* evaluate to `true`. In effect this algorithm does nothing.

  2. It does so very inefficiently: this is an `O(n^2)` algorithm and `n` is usually huge.

**Review by jonathanvdc — CHANGES_REQUESTED** [review]
The code quality in this PR is a lot better than in your previous PR. Looks like you've learned a lot in the meantime! :smile: 

Most of my comments are about style and/or design. (Which is a good thing, actually!) I did find a conceptual bug in your expatriate decoding algorithm, though. You'll want to review and fix that one as soon as possible. (i.e., before the deadline)

**Review by jonathanvdc — APPROVED** [review]
Thanks for considering my review! Your changes look great. I'm approving.

**Pull request merged**